File: CPP/program/Parkinglot.cpp
// Menggunakan pragma once agar header file hanya di-include satu kali
#pragma once

// Memasukan library yang digunakan, serta kelas yang dimiliki oleh Car dan Motorcycle yang merupakan composition
#include <iostream>
#include <string>
#include <vector>
#include "Car.cpp"
#include "Motorcycle.cpp"

// Deklarasi namespace
using namespace std;

/*
    Kelas Parkinglot yang digunakan untuk merepresentasikan data Parkinglot yang merupakan tempat parkir di luar ruangan
*/
class Parkinglot
{
    private:
        // Atribut yang digunakan pada kelas Parkinglot terdiri dari nama dan kapasitas dan jumlah kendaraan motor dan mobil.
        string nama;
        int kapasitas;
        int jumlahkendaraanMobil;
        int jumlahkendaraanMotor;
        vector<Car*> daftarMobil;
        vector<Motorcycle*> daftarMotor;
        
    public:
        // Constructor
        Parkinglot()
        {

        }

        // Constructor : Overloading
        Parkinglot(int kapasitas)
        {  
            this->kapasitas = kapasitas;
            jumlahkendaraanMobil = 0;
            jumlahkendaraanMotor = 0;
        }

        Parkinglot(int kapasitas, string nama)
        {  
            this->kapasitas = kapasitas;
            this->nama = nama;
            jumlahkendaraanMobil = 0;
            jumlahkendaraanMotor = 0;
        }

        // Method setter untuk set setiap nilai atribut pada kelas Parkinglot
        void set_kapasitas(int kapasitas)
        {
            this->kapasitas = kapasitas;
        }

        void set_nama(int nama)
        {
            this->nama = nama;
        }

        // Method untuk menambahkan motor ke dalam daftar motor di dalam class ini
        void add_motor(Motorcycle* motor) {
            if (jumlahkendaraanMotor < kapasitas)
            {
                jumlahkendaraanMotor++;
                daftarMotor.push_back(motor);
            }
            else
            {
                cout << "Maaf sudah penuh." << endl << endl;
            }
        }

        // Method untuk menambahkan mobil ke dalam daftar motor di dalam class ini
        void add_mobil(Car* mobil) {
            if (jumlahkendaraanMobil < kapasitas)
            {
                jumlahkendaraanMobil++;
                daftarMobil.push_back(mobil);
            }
            else
            {
                cout << "Maaf sudah penuh." << endl << endl;
            }
        }

        // Method getter untuk mendapatkan nilai setiap atribut pada kelas Parkinglot
        int get_kapasitas()
        {
            return kapasitas;
        }

        string get_nama()
        {
            return nama;
        }

        int get_jumlahkendaraan()
        {
            return jumlahkendaraanMobil+jumlahkendaraanMotor;
        }

        void get_datakendaraan() {
            cout << "Daftar Mobil di Parking Lot " << nama << ":" << endl << endl;
            for (Car* mobil : daftarMobil) {
                cout << "Plat Mobil " << mobil->get_platnomor() << " dengan Merk " << mobil->get_merk() << 
                " dengan Jumlah Pintu " << mobil->get_jumlahpintu() << " dengan Jumlah Kursi " << mobil->get_jumlahkursi()
                << " dirilis pada Tahun " << mobil->get_tahunproduksi() << endl;
            }
            cout << "-----------------------------------------------------------------" << endl;
            cout << "Daftar Motor di Parking Lot " << nama << ":" << endl << endl;
            for (Motorcycle* motor : daftarMotor) {
                cout << "Plat Motor " << motor->get_platnomor() << " dengan Merk " << motor->get_merk() << 
                " dengan Jenis Motor " << motor->get_jenismotor() << " dengan Kapasitas Tangki " << motor->get_kapasitastangki()
                << " dirilis pada Tahun " << motor->get_tahunproduksi() << endl;
            }
        }

        // Destructor
        ~Parkinglot()
        {

        }
};

File: Python/program/Parkinglot.py
# Deklarasi Kelas Garasi
class Parkinglot:
    __nama=""
    __kapasitasMotor=0
    __kapasitasMobil=0
    __jumlahkendaraansaatiniMotor=0
    __jumlahkendaraansaatiniMobil=0
    __daftarkendaraanMotor=[]
    __daftarkendaraanMobil=[]

    # Konstruktor kelas Garasi
    def __init__(self, nama="", kapasitasMotor=0, kapasitasMobil=0):
        self.__nama = nama
        self.__kapasitasMotor = kapasitasMotor
        self.__kapasitasMobil = kapasitasMobil
        self.__jumlahkendaraansaatiniMotor = 0
        self.__jumlahkendaraansaatiniMobil = 0
        self.__daftarkendaraanMotor=[]
        self.__daftarkendaraanMobil=[]

    # Getter dan setter nama parkiran
    def get_nama(self):
        return self.__nama
    def set_nama(self, nama):
        self.__nama = nama

    # Getter dan setter luas kapasitas
    def get_kapasitas(self):
        return self.__kapasitas
    def set_kapasitas(self, kapasitas):
        self.__kapasitas = kapasitas

    # Getter jumlah semua kendaraan di parking lot
    def get_jumlahkendaraanMotor(self):
        return self.__jumlahkendaraansaatiniMotor
    
    # Getter jumlah semua kendaraan di parking lot
    def get_jumlahkendaraanMobil(self):
        return self.__jumlahkendaraansaatiniMobil

    # Getter jumlah semua kendaraan di parking lot
    def get_jumlahkendaraanSemua(self):
        return self.__jumlahkendaraansaatiniMobil + self.__jumlahkendaraansaatiniMotor

    # Setter untuk menambah daftar kendaraan motor dan mobil
    def add_motor(self, motor):
        if self.__jumlahkendaraansaatiniMotor < self.__kapasitasMotor:
            self.__jumlahkendaraansaatiniMotor += 1
            self.__daftarkendaraanMotor.append(motor)
        else:
            print(f"\n=====\nMohon maaf, tidak bisa menambah kendaraan.\n=====\n")
    def add_mobil(self, mobil):
        if self.__jumlahkendaraansaatiniMobil < self.__kapasitasMobil:
            self.__jumlahkendaraansaatiniMobil += 1
            self.__daftarkendaraanMobil.append(mobil)
        else:
            print(f"\n=====\nMohon maaf, tidak bisa menambah kendaraan.\n=====\n")
    
    # Getter untuk menampilkan daftar kendaraan dalam garasi
    def display_kendaraan(self):
        print(f"======================================================================================================================================================")
        print(f"! Parking Lot {self.__nama} berkapasitas motor {self.__kapasitasMotor} dan berkapasitas mobil {self.__kapasitasMobil} dengan jumlah kendaraan saat ini {self.__jumlahkendaraansaatiniMobil + self.__jumlahkendaraansaatiniMotor} !\n")
        print(f"Daftar Mobil:")
        for mobil in self.__daftarkendaraanMobil:
            print(f"- Mobil dengan plat nomor {mobil.get_platnomor()} merk {mobil.get_merk()} diproduksi pada tahun {mobil.get_tahunproduksi()} berwarna {mobil.get_warna()} dengan jumlah kursi {mobil.get_jumlahkursi()} dan jumlah pintu {mobil.get_jumlahpintu()}")
        
        print(f"\nDaftar Motor:")
        for motor in self.__daftarkendaraanMotor:
            print(f"- Motor {motor.get_jenismotor()} dengan plat nomor {motor.get_platnomor()} merk {motor.get_merk()} diproduksi pada tahun {motor.get_tahunproduksi()} berwarna {motor.get_warna()} ber kapasitas {motor.get_kapasitastangki()} liter")
        print(f"======================================================================================================================================================\n")